src/ripple/legacy/0.27/CreateOffer.cpp

```diff
@@ -0,0 +1,631 @@
+                "/" << to_string (saTakerPays.getIssuer ()) <<
+                " -> " << saTakerGets.getHumanCurrency () <<
+                "/" << to_string (saTakerGets.getIssuer ());
+
+            // We use the original rate to place the offer.
+            uDirectory = getQualityIndex (uBookBase, uRate);
+
+            // Add offer to order book.
+            terResult = view.dirAdd (uBookNode, uDirectory, uLedgerIndex,
+                std::bind (
+                    &Ledger::qualityDirDescriber, std::placeholders::_1,
+                    std::placeholders::_2, saTakerPays.getCurrency (),
+                    uPaysIssuerID, saTakerGets.getCurrency (),
+                    uGetsIssuerID, uRate));
+        }
+
+        if (tesSUCCESS == terResult)
+        {
+            if (m_journal.debug)
+            {
+                m_journal.debug <<
+                    "sfAccount=" <<
+                    to_string (mTxnAccountID);
+                m_journal.debug <<
+                    "uPaysIssuerID=" <<
+                    to_string (uPaysIssuerID);
+                m_journal.debug <<
+                    "uGetsIssuerID=" <<
+                    to_string (uGetsIssuerID);
+                m_journal.debug <<
+                    "saTakerPays.isNative()=" <<
+                    saTakerPays.isNative ();
+                m_journal.debug <<
+                    "saTakerGets.isNative()=" <<
+                    saTakerGets.isNative ();
+                m_journal.debug <<
+                    "uPaysCurrency=" <<
+                    saTakerPays.getHumanCurrency ();
+                m_journal.debug <<
+                    "uGetsCurrency=" <<
+                    saTakerGets.getHumanCurrency ();
+            }
+
+            SLE::pointer sleOffer (mEngine->entryCreate (ltOFFER, uLedgerIndex));
+
+            sleOffer->setFieldAccount (sfAccount, mTxnAccountID);
+            sleOffer->setFieldU32 (sfSequence, uSequence);
+            sleOffer->setFieldH256 (sfBookDirectory, uDirectory);
+            sleOffer->setFieldAmount (sfTakerPays, saTakerPays);
+            sleOffer->setFieldAmount (sfTakerGets, saTakerGets);
+            sleOffer->setFieldU64 (sfOwnerNode, uOwnerNode);
+            sleOffer->setFieldU64 (sfBookNode, uBookNode);
+
+            if (uExpiration)
+                sleOffer->setFieldU32 (sfExpiration, uExpiration);
+
+            if (bPassive)
+                sleOffer->setFlag (lsfPassive);
+
+            if (bSell)
+                sleOffer->setFlag (lsfSell);
+
+            if (m_journal.debug) m_journal.debug <<
+                "final terResult=" << transToken (terResult) <<
+                " sleOffer=" << sleOffer->getJson (0);
+        }
+    }
+
+    if (terResult != tesSUCCESS)
+    {
+        m_journal.debug <<
+            "final terResult=" << transToken (terResult);
+    }
+
+    return terResult;
+}
+
+//------------------------------------------------------------------------------
+
+std::pair <bool, TER>
```

Comment:
That return type needs some better documentation, I still haven't really managed to work out what the `bool` means!